apps/captain_fact/lib/captain_fact/videos/captions_fetcher.ex

```diff
@@ -0,0 +1,40 @@
+defmodule CaptainFact.Videos.CaptionsFetcher do
+  @moduledoc """
+  Fetch captions for videos.
+  """
+
+  if Application.get_env(:db, :env) == :test do
```

Comment:
This is not a good practice, but no judgement as I've done this at other parts of the code :imp: The recommended way for this kind of situations in Elixir is to define a `CaptionFetcher` [behaviour](https://elixir-lang.org/getting-started/typespecs-and-behaviours.html), then to have two implementations (`CaptionsFetcherYoutube` + `CaptionsFetcherTest`) and to alternate between the two by setting the fetcher implementation in config.

This avoid having test-specific code in modules.

Do not bother to change this if you don't want to, as I said I've used this trick too in parts of the code :wink:

Reply:
Thanks! Do you have an example in the code base using the behaviour?
I do understand the reasoning but when I saw you using the `if Application.get_env....` I thought that's actually an easy solution :-)

Reply:
I don't have any examples in this project, but it would look like:

## Implementation

`lib/captain_fact/videos/captions_fetcher.ex`
```elixir
defmodule CaptainFact.Videos.CaptionsFetcher do
  @moduledoc """
  A behaiour for fetching captions.
  """

  @callback fetch(string()) :: {:ok, DB.Schema.Caption.t()} | {:error, binary()}
end
```

`lib/captain_fact/videos/captions_fetcher_test.ex`
```elixir
defmodule CaptainFact.Videos.CaptionsFetcherTest do
  @moduledoc """
  A mock for faking captions fetching requests.
  """

  @behaviour CaptainFact.Videos.CaptionsFetcher

  @impl true
  def fetch(provider_id) do
    %DB.Schema.Caption{content: "TEST"}
  end
end
```

`lib/captain_fact/videos/captions_fetcher_youtube.ex`
```elixir
defmodule CaptainFact.Videos.CaptionsFetcherYoutube do
  @moduledoc """
  A captions fetcher for YouTube.
  """

  @behaviour CaptainFact.Videos.CaptionsFetcher

  @impl true
  def fetch(youtube_id) do
    with {:ok, content} <- fetch_captions_content(youtube_id, locale) do
      {:ok, %{content: content, format: "xml", locale: locale, is_auto_generated: true}}
    end
  end
end
```

## Configuration

Then in `apps/captain_fact/config/config.exs`:

```elixir
config :captain_fact,
  captions_fetcher: CaptainFact.Videos.CaptionsFetcherYoutube
```

In `apps/captain_fact/config/test.exs`:
```
config :captain_fact,
  captions_fetcher: CaptainFact.Videos.CaptionsFetcherTest
```

## Usage

And finally, in `captain_fact/videos/videos.ex`:

```elixir
defmodule CaptainFact.Videos do
  @moduledoc """
  The boundary for the Videos system.
  """

  import Ecto.Query, warn: false
  import CaptainFact.Videos.MetadataFetcher
  import CaptainFact.Videos.CaptionsFetcher
  # ...
  alias CaptainFact.Accounts.UserPermissions

  @captions_fetcher Application.get_env(:captain_fact, :captions_fetcher) # <=====

  # ...

  def download_captions(video = %Video{}, locale) do
  with {:ok, captions} <- @captions_fetcher.fetch({video.provider, video.provider_id}, locale) do
    ...
```
```